### PR TITLE
S3 auth: Fix malformed `Authorization` header

### LINKED ISSUE #58643: S3 Authentication: Malformed `Authorization` HTTP header
### What is the bug or the crash?

When adding a remote XYZ layer using the AWS S3 protocol, QGIS is expected to automatically insert certain HTTP headers to authenticate requests it makes to S3:
- `Authorization`
- `X-Amz-Content-SHA256`
- `X-Amz-Date`

_Note: This is probably true for other layer types besides XYZ, but I haven't tested with all layer types._

The issue is specifically with the format of the auto-generated `Authorization` header value.

The `Authorization` string format expected by AWS is:

```
AWS4-HMAC-SHA256 Credential=<redacted>, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=<redacted>
```
Reference: https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-authenticating-requests.html

The `Authorization` string format generated by QGIS is:

```
AWS4-HMAC-SHA256Credential=<redacted>, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=<redacted>
```

Note the missing space before `Credential`.

Because of this malformed `Authorization` header, all QGIS requests to the S3 bucket fail with status code 400 and the following payload:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<Error>
    <Code>InvalidArgument</Code>
    <Message>Unsupported Authorization Type</Message>
    <ArgumentName>Authorization</ArgumentName>
    <ArgumentValue>AWS4-HMAC-SHA256Credential=<redacted>, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=<redacted></ArgumentValue>
    <RequestId>Z88ASX2YKJ46M185</RequestId>
<HostId>B//KAO3Poe330fIwR2RtwfR4+7OkPFzK0GEUhTWrdBd3nRsG44gOKlTlqWARLqgS//DtZL5kh62YjbiGaM4XlrSlcq3JMe96</HostId>
</Error>
```

If I reproduce the same request QGIS makes to S3 in Postman, and add the missing space before `Credential`, the request succeeds.

### Steps to reproduce the issue

1. Open a new project in QGIS Desktop
2. Select "Layer" > "Add Layer" > "Add XYZ Layer...".
image

3. Add an S3 "Layer Source" URL, e.g. `https://s3.ca-central-1.amazonaws.com/my-bucket/{z}/{x}/{y}.png`
image

4. Under "Authentication", click the "Create new configuration" button. Select the "AWS S3" configuration type and enter your AWS credentials as such, then click "Save".
image

5. Click "Add" to add the new XYZ layer.
image

6. Open the Debugging/Development Tools panel ("View" > "Panels" > "Debugging/Development Tools"), then click the "Record logs" button.

7. You should see failed (red) requests to your S3 bucket appear in the "Debugging/Development Tools" panel as you interact with the map in QGIS.

image

8. Expand one of the failed S3 requests to inspect it. Notice the response status code (400), and the unexpected `Authorization` request header format (no space before `Credential=`):

image

9. Replicate the same request in Postman. You should get the following error response from AWS:

image

11. Fix the `Authorization` header format in Postman by adding the missing space before `Credential=`. Execute the request again and notice it succeeds (200 status code).

image

### Versions

QGIS version | 3.38.1-Grenoble | QGIS code revision | 3d4177afc6b
-- | -- | -- | --
Qt version | 5.15.2
Python version | 3.9.5
GDAL/OGR version | 3.3.2
PROJ version | 8.1.1
EPSG Registry database version | v10.028 (2021-07-07)
GEOS version | 3.9.1-CAPI-1.14.2
SQLite version | 3.35.2
PDAL version | 2.3.0
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.11.5
OS version | macOS 14.6
  |   |   |  
Active Python plugins
quick_map_services | 0.19.34
SentinelHub | 2.0.2
qtiles | 1.7.2
processing | 2.12.99
grassprovider | 2.12.99
db_manager | 0.1.20
MetaSearch | 0.3.6

</body></html>QGIS version
3.38.1-Grenoble
QGIS code revision
[3d4177afc6b](https://github.com/qgis/QGIS/commit/3d4177afc6b)
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.3.2
PROJ version
8.1.1
EPSG Registry database version
v10.028 (2021-07-07)
GEOS version
3.9.1-CAPI-1.14.2
SQLite version
3.35.2
PDAL version
2.3.0
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.11.5
OS version
macOS 14.6

Active Python plugins
quick_map_services
0.19.34
SentinelHub
2.0.2
qtiles
1.7.2
processing
2.12.99
grassprovider
2.12.99
db_manager
0.1.20
MetaSearch
0.3.6

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

It appears the S3 `Authorization` header format could be fixed in the `qgsauthawss3method.cpp` file here: https://github.com/qgis/QGIS/blob/master/src/auth/awss3/core/qgsauthawss3method.cpp#L132

I'm currently attempting to build QGIS locally for the first time to confirm that editing that `qgsauthawss3method.cpp` file indeed fixes the issue.

### PR DESCRIPTION
**Pull request by pascal-giguere**
## Description

Fixes https://github.com/qgis/QGIS/issues/58643

This change fixes request authentication when loading a layer from a private S3 bucket by correcting the format of the `Authorization` header.

____

**Before**

The malformed `Authorization` header (note the missing space before `Credential=`) leads to failed requests (400 status code).

image

____

**After**

image

The corrected `Authorization` header leads to successful requests (200 status code).

____

I tested the change by doing a local build on Mac, and repeating the reproduction steps described in https://github.com/qgis/QGIS/issues/58643.

### CODE REVIEW

File: src/auth/awss3/core/qgsauthawss3method.cpp [hunk 1]
```diff
@@ -129,7 +129,7 @@ bool QgsAuthAwsS3Method::updateNetworkRequest( QNetworkRequest &request, const Q
   request.setRawHeader( QByteArray( "Host" ), request.url().host().toLocal8Bit() );
   request.setRawHeader( QByteArray( "X-Amz-Date" ), dateTime );
   request.setRawHeader( QByteArray( "Authorization" ),
-                        encryptionMethod + "Credential=" + username + '/' + date + "/" + region + "/s3/aws4_request, SignedHeaders=" + headerList + ", Signature=" + signature );
+                        encryptionMethod + " Credential=" + username + '/' + date + "/" + region + "/s3/aws4_request, SignedHeaders=" + headerList + ", Signature=" + signature );
```
Comment (3nids) [inline src/auth/awss3/core/qgsauthawss3method.cpp:132]: would it be clearer with QString::arg ?

Comment (pascal-giguere) [reply src/auth/awss3/core/qgsauthawss3method.cpp]: I think it would yeah, just refactored it!

File: src/auth/awss3/core/qgsauthawss3method.cpp [hunk 2]
```diff
@@ -125,11 +125,13 @@ bool QgsAuthAwsS3Method::updateNetworkRequest( QNetworkRequest &request, const Q
                                 QCryptographicHash::Sha256 );
 
   const QByteArray signature = QMessageAuthenticationCode::hash( stringToSign, signingKey, QCryptographicHash::Sha256 ).toHex();
+  const QByteArray authorizationHeader = QString("%1 Credential=%2/%3/%4/s3/aws4_request, SignedHeaders=%5, Signature=%6")
+                                             .arg(encryptionMethod, username, date, region, headerList, signature)
+                                             .toLocal8Bit();
```
Comment (agiudiceandrea) [inline src/auth/awss3/core/qgsauthawss3method.cpp:130]: Is it safe to use `toLocal8Bit()` here (and in some of the various other occurrences)? Shouldn't `toUtf8`() be used instead?

Comment (rouault) [reply src/auth/awss3/core/qgsauthawss3method.cpp:130]: yes toUtf8() seems far better/safer

Comment (pascal-giguere) [reply src/auth/awss3/core/qgsauthawss3method.cpp]: I updated all occurrences to use `toUtf8()`

**Review by 3nids — APPROVED** [review]

**Comment by Djedouas** [comment]
Thanks! Looks good to me 😃

**Pull request merged**